### ISSUE TITLE
Gradle 6.4-rc-1 triggers compileJava when project version is changed in multi-module project

### ISSUE THREAD

**Issue by asodja** [gradle/gradle#12860]
So I noticed that with Gradle 6.4-rc-1 when you change project version compileJava is not taken from build cache even if all files were compiled before. I made a simple reproducer: https://github.com/asodja/gradle-multi-project-cache. You will notice this change in behavior:

Gradle 6.3:

    gradle clean compileJava -> first run, compileJava is NOT taken from build cache
    gradle clean compileJava -> second run, compileJava is taken from build cache

Gradle 6.4-rc-1:

    gradle clean compileJava -> first run, compileJava is NOT taken from build cache
    gradle clean compileJava -> second run, compileJava is NOT taken from build cache

Caching is enabled with `org.gradle.caching=true` in `gradle.properties`.

In this reproducer I force project version change on every build with: `version = UUID.randomUUID().toString()` in build.gradle.

I use `implementation project(":<otherproject>")` to set dependencies. 

---
Possible cases where this might be an issue: 
- This might add extra time when version is changed on release builds (for example changing version from 1.0.0-SNAPSHOT to 1.0.0), even though previous CI/CD builds passed
- This might impact compile time when you switch branches that are on different versions etc.

Is this a desired behaviour or am I doing something wrong? If you have a big multi-module project, just a project version change might add a lot of undesired extra compile time.

---
cc: @gradle/build-cache

**Comment by asodja** [comment]
Ok, first sorry for using wrong labels. I am obviously not a member, just not fluent with your Github issues and used wrong category.

Second, some extra information regarding this. From ` gradle clean compileJava --info` it looks like options.javaModuleVersion triggers this:
```
Build cache key for task ':library1:compileJava' is a58ac2060bc2663e4b1a90802b090dfa
Task ':library1:compileJava' is not up-to-date because:
  Value of input property 'options.javaModuleVersion' has changed for task ':library1:compileJava'
The input changes require a full rebuild for incremental task ':library1:compileJava'.
Full recompilation is required because no incremental change information is available. This is usually caused by clean builds or changing compiler arguments.
```

**Comment by jjohannes** [comment]
Thanks for reporting this @asodja! It's a regression caused by a new default behavior for the Java module support we added. It's fixed for RC2 through #12863.
